src/crewai/agent.py

```diff
@@ -235,6 +241,28 @@ def _setup_agent_executor(self):
             self.cache_handler = CacheHandler()
         self.set_cache_handler(self.cache_handler)
 
+    def _set_knowledge(self):
+        try:
+            if self.knowledge:
+                knowledge_agent_name = f"{self.role.replace(' ', '_')}"
+                if isinstance(self.knowledge, dict):
+                    knowledge_data = self.knowledge.copy()
+                    knowledge_data["store_dir"] = knowledge_agent_name
+                    self.knowledge = Knowledge(**knowledge_data)
+                    self.knowledge.storage.initialize_knowledge_storage()
```

Comment:
Shouldn't this be handled in the constructor of Knowledge.py?

Why does the agent need to know to call initialize_knowledge_storage?

Reply:
There’s cases where it initializes before agent is done initializing.